Fix Floating Origin popping

daleeidd opened the pull request:
_**Status:** Improvement. But not complete._

Fixes Floating Origin popping on teleports for:
- mesh
- normal maps
- foam
- bubbles
- caustics
- persistent simulations

But the following is still a problem:
- [x] wave jitter (very noticeable in whirlpool scene and might have something to do with combine shader)
- [x]  collision queries?
- [x] conditional compilation
- [x] wave patch

daleeidd commented:
@huwb any thoughts on this so far?

huwb commented:
Yeah looks good so far.

Yeah i suppose the wave queries need the offset too

Ideally the offset would be mod'd down. The whole point of floating origin is to take away a large amount of the position, i.e. if player is 10km from origin, an offset of 10km is subtracted from everything. You store 10km in a variable and then add it everywhere to compensate and i think we unfortunately expose ourselves to the original precision issues.

For a normal map that tiles every say 12m, it should be possible to instead of storing the full world space offset, store `offset % 12f` which will be much smaller. that might be a good option for things that obviously repeat like textures, assuming we can read their periodicity from the material params or etc.

for waves its a bit trickier. maybe modding with the size of the biggest wave slice does it. maybe modding with 2048 is big enough for 99% of situations.

does this make sense?

daleeidd commented:
> Ideally the offset would be mod'd down. The whole point of floating origin is to take away a large amount of the position, i.e. if player is 10km from origin, an offset of 10km is subtracted from everything. You store 10km in a variable and then add it everywhere to compensate and i think we unfortunately expose ourselves to the original precision issues.

Thanks. I didn't see any anomalies without it, but make sense.

> For a normal map that tiles every say 12m, it should be possible to instead of storing the full world space offset, store `offset % 12f` which will be much smaller. that might be a good option for things that obviously repeat like textures, assuming we can read their periodicity from the material params or etc.

`% _NormalsScale * 2.0 * lodDataGridSize` appeared to work in all cases. We could either use a keyword or calculate in C# and pass through to each chunk. Will look at doing the other textures.

> for waves its a bit trickier. maybe modding with the size of the biggest wave slice does it. maybe modding with 2048 is big enough for 99% of situations. so basically LodDataMgrAnimWaves could store `offset % 2048` as a test. Maybe this works for other loddata too.

In regards to the jitter? It appears to be something else as it occurs even not far from zero. It looks like it has something to do with scale, LOD count, and FO threshold. Eyeballing the data there is a slight difference between non FO and FO:

<img width="384" alt="111" src="https://user-images.githubusercontent.com/5249806/146169958-1ae569b8-fbc9-440a-aaf6-c50f71319023.png">

That being said, I was able to reduce (possibly eliminate) the jitter by playing with LOD scale, LOD count, and LOD resolution (and in some case FO threshold), but I haven't been able to determine how these values correlate mathematically yet. Perplexed why this is even an issue at like 100 m from zero.

huwb commented:
Hmm weird. 100m from zero presumably has a very small offset - but adding this offset makes it jitter?

Test sounds good

daleeidd commented:
Done all of the textures and sims using mod.

> Hmm weird. 100m from zero presumably has a very small offset - but adding this offset makes it jitter?

Yeah. It shifts it a little bit when the origin is moved. I thought I managed to fix it but it's still there.

> Yeah i suppose the wave queries need the offset too

Using the query visualiser it is possible to add the offset as the queries return and keep the result in the expected place. But not sure whether this is the correct approach yet. Anything running in FixedUpdate probably won't be able to use this approach as FO is LateUpdate.

Will put this down for a little bit and push other stuff forward.

daleeidd commented:
I think this is ready for early review...

I moved FO to FixedUpdate which solves queries I think. If queries return everything in local space then I don't think we need to add FO offset to it. Is that correct?

There are two dev scenes. One is for LOD data and the other for Ocean.shader textures. Once #931 I will merge them together using some of the scripts from there and prefab it.

For the first scene, just hit play and it will cross the threshold and should pause automatically.

For the second scene, before entering play mode, hit pause and then enter play mode so you are in paused state, then you can step through the frames and cross the threshold. The camera moves automatically and threshold is less than a unity away.

I added some general debug options and developer only debug options:
![1231312](https://user-images.githubusercontent.com/5249806/151645010-172d8e87-d580-41a2-bd6c-5d8d38e640b8.jpg)
Anything under "Developer Only" only appears if CREST_DEBUG symbol is present. Those debug options are for positioning the camera at an offset from the threshold.

As for the waves have a slight pop, I think I have fixed it. There is still a pixel level pop but I doubt it is noticeable. I am viewing things with the ocean in a paused state but in movement probably won't be noticeable.

I'll need to add release notes still and clean up.

daleeidd commented:
I found that the combination of restricting the threshold to a minimum of 4096 and power of 2 to avoid wave popping. Also I only shift the origin by threshold sized increments which also was needed. Aforementioned is done in FloatingOrigin.cs. With this subtracting the origin in the wave shader has no positive effect and probably can be removed. I will have to investigate further but wouldn't mind hearing your thoughts.

huwb commented on crest/Assets/Crest/Crest/Scripts/Helpers/FloatingOrigin.cs:
```diff
@@ -18,7 +18,9 @@
 
+using System.Collections;
 using UnityEngine;
+using UnityEngine.Experimental.Rendering;
```
does this work in our current supported versions? just making sure there isnt something that got moved out of Experimental

daleeidd replied:
It is experimental for all versions so far. I can surround it and dependent code with CREST_DEBUG for now? It is for the RenderDoc capture which I doubt anyone will need.

daleeidd replied:
After reading your other comment I will wrap with CREST_DEBUG as going to wrap that stuff too.

daleeidd replied:
Done. I have wrapped it with CREST_DEBUG and wrapped RenderDoc too. Thanks.

huwb commented on crest/Assets/Crest/Crest/Scripts/Helpers/FloatingOrigin.cs:
```diff
@@ -65,35 +69,149 @@ public class FloatingOrigin : MonoBehaviour
         [Tooltip("Optionally provide a list of Gerstner components to avoid doing a FindObjectsOfType() call."), SerializeField]
         ShapeGerstnerBatched[] _overrideGerstnerList = null;
 
+        [Space(10)]
```
this is good. i should probably look at adding spaces to try make OceanRenderer ui a bit clearer

huwb commented on crest/Assets/Crest/Crest/Scripts/Helpers/FloatingOrigin.cs:
```diff
@@ -65,35 +69,149 @@ public class FloatingOrigin : MonoBehaviour
+
+        void FixedUpdate()
+        {
+            if (!_isFirstFixedUpdate)
```
the way i've typically handled this is something like

int _lastUpdateFrame;

void FixedUpdate()
{
    if( _lastUpdateFrame == Time.frameCount )
    {
        return;
    }
    _lastUpdateFrame = Time.frameCount;

    ....
}

this would be a bit simpler for me. also if this were to mean we could make the coroutine above debug-only (so not in final builds) i'd be in favour - but only because i'm hopelessly bad at coroutines and view them with suspicion :). but if you prefer to keep it the way it is and the coroutine is not generating per frame garbage then i wouldnt block the review on it.

daleeidd replied:
Thanks. I have changed to your approach as it does reduce our dependency on the coroutine. But we still depend on it because TeleportOriginThisFrame and HasTeleportedThisFrame has to be cleared but FixedUpdate does not run each frame.

daleeidd replied:
As for garbage, I did not see any from this. Found some other garbage :\ which I will fix separately.

huwb commented on crest/Assets/Crest/Crest/Scripts/Helpers/FloatingOrigin.cs:
```diff
@@ -65,35 +69,149 @@ public class FloatingOrigin : MonoBehaviour
+
+                if (_debug._isCapturing)
+                {
+                    ExternalGPUProfiler.EndGPUCapture();
```
interesting, nice

daleeidd replied:
Indeed. It was very useful/necessary as LOD doesn't run when paused.

huwb commented on crest/Assets/Crest/Crest/Scripts/Helpers/FloatingOrigin.cs:
```diff
@@ -65,35 +69,149 @@ public class FloatingOrigin : MonoBehaviour
+        void OnValidate()
+        {
+            // Must be power of two to avoid popping.
+            _threshold = Mathf.Pow(2f, Mathf.Round(Mathf.Log(_threshold, 2f)));
```
does > 2048 need to be validated here too? I know you have a min attribute but its public and could be set from code or users could hack the min i guess

daleeidd replied:
Good find. Done. Thanks.

huwb commented on crest/Assets/Crest/Crest/Shaders/OceanConstants.hlsl:
```diff
@@ -13,6 +13,9 @@
 // C# code to check if any parameters are within the MAX_LOD_COUNT limits
 #define MAX_LOD_COUNT 15
 
+// The size of the largest LOD.
```
is this the size of the largest lod? i think it might be good to expand this comment and rename the symbol to make it really explicit and clear. i cant quite remember the background for this, i guess its in our thread but it would be good to make the code really clear

daleeidd replied:
I have since removed it as the issue was actually a bug in C#.

huwb commented on crest/Assets/Crest/Crest/Shaders/OceanEmission.hlsl:
```diff
@@ -70,7 +70,8 @@ void ApplyCaustics
 	// could sample from the screen space shadow texture to attenuate this..
 	// underwater caustics - dedicated to P
-	const float3 scenePosUV = WorldToUV(i_scenePos.xz, cascadeData1, _LD_SliceIndex + 1);
+	// TODO: Investigate if necessary anymore.
```
soft vote for looking into this if its fast rather than adding a todo, but also fine with it being deferred.

daleeidd replied:
I've removed it. Wasn't necessary anymore after testing.

huwb commented on crest/ProjectSettings/ProjectSettings.asset:
```diff
@@ -620,7 +620,7 @@ PlayerSettings:
   webGLDecompressionFallback: 0
   scriptingDefineSymbols:
-    1: CROSS_PLATFORM_INPUT;UNITY_POST_PROCESSING_STACK_V2;CREST_OCEAN
+    1: CROSS_PLATFORM_INPUT;UNITY_POST_PROCESSING_STACK_V2;CREST_OCEAN;CREST_DEBUG
```
just double checking this is intentional? not objecting just checking

daleeidd replied:
It is. For convenience since FO testing is wrapped with this.

huwb commented on crest/Assets/Crest/Crest/Shaders/OceanEmission.hlsl:
```diff
@@ -70,7 +70,8 @@ void ApplyCaustics
 	// underwater caustics - dedicated to P
-	const float3 scenePosUV = WorldToUV(i_scenePos.xz, cascadeData1, _LD_SliceIndex + 1);
+	// TODO: Investigate if necessary anymore.
+	const float3 scenePosUV = WorldToUV(CREST_WITH_FLOATING_ORIGIN_LOD_OFFSET_XZ(i_scenePos.xz), cascadeData1, _LD_SliceIndex + 1);
```
this is the only case where this FO macros is used going into WorldToUv, it might be good to document why its needed here. or maybe thats the TODO?

daleeidd replied:
No longer needed. Fixed in C#. Thanks :^)

huwb commented on crest/Assets/Crest/Crest/Shaders/OceanHelpersNew.hlsl:
```diff
@@ -21,6 +21,18 @@ float3 WorldToUV(in float2 i_samplePos, in CascadeParams i_cascadeParams, in flo
 	return float3(uv, i_sliceIndex);
 }
 
+float3 WorldToUVSource(in float2 i_samplePos, in CascadeParams i_cascadeParams, in float i_sliceIndex)
```
i didnt automatically understand the reason for this function. is there an iteration on the naming that would make it clearer (even if its quite explicit), or could we add a comment above it. just to make it clear which function to use from this library for different situations.

daleeidd replied:
Removed as fixed in C#.

huwb reviewed (commented):
Thats it for this pass, awesome work. Thanks for getting all the debug functionality in!

daleeidd commented:
Feedback addressed. Thanks.

I was able to remove most of the shader code as I was able to fix it in C# instead.

After review I will fix up the dev stuff and do release notes / docs. But I think this fixes everything FO wise.

huwb commented on crest/Assets/Crest/Crest/Shaders/Resources/UpdateFoam.compute:
```diff
@@ -41,30 +41,33 @@ void UpdateFoam(uint3 id : SV_DispatchThreadID)
 	const CascadeParams cascadeData = _CrestCascadeData[sliceIndex];
 	const float2 worldPosXZ = UVToWorld(input_uv, sliceIndex, cascadeData);
 	const float3 uv_slice = float3(input_uv, id.z);
+	half foam = 0.0;
```
the diff of this file is a little hard to review here which is fine but just checking this is safe to do? the logic below can be a bit delicate but if it looks good ignore this comment

daleeidd replied:
It is only whitespace changes sorry. If hide whitespace is enabled it makes it easier:

![1_1](https://user-images.githubusercontent.com/5249806/153779615-5ca55ab6-eeb6-4505-bb3e-f0c52a71cd29.jpg)

daleeidd replied:
In case you're on mobile and it isn't available:

![22](https://user-images.githubusercontent.com/5249806/153779793-3032bd15-ebcd-4b53-8e1a-974cab49ae34.jpg)

I might drop the whitespace changes and do separately.

huwb replied:
Oh forgot about that thanks. Looks good to me, feel free to keep

huwb reviewed (approved):
Yeah just one paranoia question but ill approve, nice one

daleeidd commented:
I've added release notes and removed paragraphs about FO limitations since I believe there aren't any anymore 🤞

huwb commented:
Great!

nitroman123 commented:
Everything work well I've tested , no popping not nothing well done !!!
